src/main/scala/com/fulcrumgenomics/bam/FindTechnicalReads.scala

```diff
@@ -77,33 +93,66 @@ class FindTechnicalReads
       val r1 = iter.next()
       val r2 = if (iter.hasNext && iter.head.getReadName == r1.getReadName) Some(iter.next()) else None
       val recs = Seq(Some(r1), r2).flatten
-      if (recs.exists(r => matcher.matches(r.getReadBases))) {
-        recs.foreach(out.addAlignment)
-        n += recs.size
+
+      val isTechnical = if (isUnmappedTemplate(r1)) {
+        val hit      = recs.toStream.flatMap(r => matcher.findMatch(r.getReadBases)).headOption
+        val hitIndex = hit.map(h => sequenceToIndex(h))
+
+        for (t <- tag; hi <- hitIndex; r <- recs) {
+          r.setAttribute(t, hi)
+          n += 1
+        }
+
+        hit.isDefined
+      }
+      else {
+        false
       }
 
+      if (isTechnical || allReads) recs.foreach(out.addAlignment)
       recs.foreach(progress.record)
     }
 
-    logger.info(s"Wrote out ${n} reads that matched to technical sequences.")
+    logger.info(s"Found ${n} reads that matched to technical sequences.")
```

Comment:
I have a strong preference for using `${}` over bar `$` in these strings - I think it makes it both much easier to spot where the interpolation is happening and also to adjust if necessary, e.g. `${n}` -> `${n-1}`.